### ISSUE TITLE
Doesn't apppear to support bidirectional / net meters

### ISSUE THREAD

**Issue by jeremyakers** [tedpearson/electric-usage-downloader#7]
Hello there,

This project is amazing. Thank you so much for publishing this. I've been working (on and off) trying to reverse engineer the JSON response that SmartHub sends back to my browser so that I could automate something to grab the data daily and stuff it into a database for visualization.

The JSON response is so large I've always just given up after looking at it for a few minutes. This project helped me narrow in on exactly which part of the JSON response carries the actual consumption data... however as I look at the code and compare it to the JSON data I'm wondering if this may return incorrect data for some use cases. In my case, for example, I have a Net meter with Solar panels so when I go into SmartHub there are actually 3 consumption numbers: Consumption, Generation, and "Net". So for example let's say we're looking at say 9am and the sun is just barely up. During most of that hour I'm probably exporting a small amount of power, let's say 0.5 kWh, but then at 9:45 my wife turns on her hair dryer and we end up pulling 0.1kWh from the grid before 10am. The numbers for that would be:

Generation: 0.5 kWh
Consumption: 0.1 kWh
Net -0.4 kWh

I don't have an InfluxDB set up to test this code as is, but I've commented out the InfluxDB portion and have been just printing out the values and it it appears to me that it's hard coded to "Series[0]" in the code and when I look at the JSON document I get back in Postman when I call the API directly, I get TWO series arrays under usage:

```
                "series": [
                    {
                        "type": "column",
                        "color": "#00BB33",
                        "name": "658973 - Consumption",
                        "meterNumber": "658973",
                        "data": [
                            {
                                "x": 1713398400000,
                                "y": 0.43,
                                "enableDrilldown": true
                            },
                            <series data continues>
                    },
                    {
                        "type": "column",
                        "color": "#FF5319",
                        "name": "658973 - Generation",
                        "meterNumber": "658973",
                        "data": [
                            {
                                "x": 1713398400000,
                                "y": 0.0,
                                "enableDrilldown": true
                            },
```

If I'm understanding this code correctly; I think this will give me my "consumption" numbers only, but not my generation or my "Net" values.

Anyway: I just thought I would share my findings in case you're interested in having this code support solar / net meters. This has certainly given me what I need to get the data for myself so I deeply appreciate you publishing this!

**Comment by tedpearson** [comment]
I'm glad this has helped you out with your project!

As for the series data, I hard coded it to use the first entry because my data only returns a single series without any information to suggest what another series might be. Comparing your json to mine, the only difference in those values is in the `name` field. If that's the only difference other than the multiple series, I would probably add a tag to the metrics using the text part of the series name if there is more than one series.

Here's the full output I get for my single series, maybe there are more differences in another field between your net metering setup and mine?

  
                "series": [
                    {
                        "type": "column",
                        "color": "#00BB33",
                        "name": "350012589",
                        "meterNumber": "350012589",
                        "data": [
                            {
                                "x": 1713312000000,
                                "y": 0.64,
                                "enableDrilldown": true
                            },
                            <series data continues>
                        "isNet": false,
                        "stack": "non-net-meter",
                        "dataGrouping": {
                            "enabled": false,
                            "smoothed": false
                        },
                        "yAxis": 0,
                        "zIndex": 98,
                        "visible": true,
                        "turboThreshold": 0,
                        "channel": 1
                    }
                ],

**Comment by jeremyakers** [comment]
It appears I may have trimmed too much from my JSON example above, let me try again:

```
                "series": [
                    {
                        "type": "column",
                        "color": "#00BB33",
                        "name": "658973 - Consumption",
                        "meterNumber": "658973",
                        "data": [
                            {
                                "x": 1713398400000,
                                "y": 0.43,
                                "enableDrilldown": true
                            },
                            <series data continues>
                        ],
                        "isNet": false,
                        "stack": "non-net-meter",
                        "dataGrouping": {
                            "enabled": false,
                            "smoothed": false
                        },
                        "yAxis": 0,
                        "zIndex": 96,
                        "visible": true,
                        "turboThreshold": 0,
                        "channel": 1
                    },
                    {
                        "type": "column",
                        "color": "#FF5319",
                        "name": "658973 - Generation",
                        "meterNumber": "658973",
                        "data": [
                            {
                                "x": 1713398400000,
                                "y": 0.0,
                                "enableDrilldown": true
                            },
                            <series data continues>
                        ],
                        "isNet": false,
                        "stack": "non-net-meter",
                        "dataGrouping": {
                            "enabled": false,
                            "smoothed": false
                        },
                        "yAxis": 0,
                        "zIndex": 98,
                        "visible": true,
                        "turboThreshold": 0,
                        "channel": 2
```

So yeah, it appears the only difference is the name. I didn't even notice the fact that mine actuall says:
```
                        "isNet": false,
                        "stack": "non-net-meter",
```

Maybe because the *type* of net meter I have is basically just two meters in one. Perhaps there are other types of net meters that behave differently?

**Comment by tedpearson** [comment]
Interesting! I have one more question: the "ELECTRIC" array has two entries in my case, one with `"type": "COST"` and one with `"type": "USAGE"`. In my case, both have a `series` array with a single element. Does your data have two `series` items for both? I'm also wondering if so, if the Generation "COST" series is negative or positive. Or maybe it's not even called "COST".

    {
        "status": "COMPLETE",
        "data": {
            "ELECTRIC": [
                {
                    "series": [
                        {
                            "data": [
                                {
                                    "x": 1713312000000,
                                    "y": 0.64,
                                    "enableDrilldown": true
                                },
                                ...
                            ],
                            ...
                        }
                    ],
                    "type": "USAGE"
                },
                {
                    "series": [
                        {
                            "data": [
                                {
                                    "x": 1713312000000,
                                    "y": 0.06,
                                    "enableDrilldown": true
                                },
                                ...
                            ],
                            ...
                        }
                    ],
                    "type": "COST"
                }
            ]
        }
    }

**Comment by jeremyakers** [comment]
I could send you a copy of my response JSON if you like. I don't really want to post it here since it has some personal info in it, but I can PM or email you.

**Comment by jeremyakers** [comment]
I don't see any ways to contact you on your GitHub profile. Just let me know if you have a preferred method of me just sending over my JSON file. If you want: I have my twitter handle and Mastodon handle on my Github profile so if you want to shoot me an email address or something you can PM it to me on one of those.

**Comment by tedpearson** [comment]
@jeremyakers I don't know if you have any interest in looking over the fix before I merge it in, but it's available in #9.

**Comment by tedpearson** [comment]
For future reference, the data did have two series for "COST" as well as for "USAGE". All values in "COST" were negative for the "Generation" series.

Also, the "meters" field had `"flowDirection": "REVERSE"` and `"isNetMeter": true` while mine had `"flowDirection": "TOTAL"` and `"isNetMeter": false`

**Comment by jeremyakers** [comment]
Not related to this issue... But something else I thought I'd mention. As I was playing with the API I was able to pull more than 45 days worth of data when invoking via Postman. I didn't spend too much time testing, but I was able to pull 90 days successfully and when I jumped to 120 it still returned a JSON response... no error, but the JSON was mostly empty (No usage or cost arrays, etc) so I'm guessing that was just too much data for it to process. So the 45 day limit may be different depending on who your provider is.

Along the same lines: I was able to go back and pull data from prior years as well. I went back and pulled data from Jan 2023 by giving it a start date of midnight Jan 1 2023 and end date of midnight Feb 1 2023. So if anyone is interested in going back and pulling historical data you can seemingly do that in 30-90 day chunks. I was able to pull all of my 2023 data one month at a time.

**Comment by tedpearson** [comment]
@jeremyakers Interestingly, when i tested my coop just now, I was able to pull up to 365 days of data at a time (I didn't test past that threshold), and it looked to be correct from a spot check. But I will note that if I try to do this via the Novec Smarthub, it won't let me exceed 30 days: 
<img width="611" alt="image" src="https://github.com/tedpearson/electric-usage-downloader/assets/1477390/4fcb1660-265a-4ee4-92b7-fa860790ef6f">

So I'll leave the limits alone and leave modifying the code to allow longer date ranges to the advanced user. 🙂 

I also pulled back data, though 45 days at a time. 😉 I was able to go back to April 2022, which I think is around when my smart meter was installed. This is the reason that I provide the `--start yyyy-mm-dd --end yyyy-mm-dd` arguments to the program.
